### PR TITLE
Add support for pre-commit tool.

### PR DESCRIPTION
**Pull request by fkouteib**
#### Problem
* Chore code hygiene checks (i.e. trim trailing whitespace) not covered by rust fmt, and generally require manual config in devs' IDEs. 
* No guardrails to prevent accidentally committing secrets to the repo.

#### Summary of Changes
* add support for automated checks and non-functional fixes not covered by cargo for general code hygiene (i.e. whitespace; EOF; preventing large file check-in).
* Add support for automated gitleaks checks to prevent committing secrets to repo.
* These checks would run by default on every git commit invokation by a dev. Enforcement added in CI to catch any misses in PR.
* The first invokation can take a couple minutes to install the hooks. All future invokations will only run on modified files by default making it really fast.

### CODE REVIEW

**Comment by bw-solana** [comment]
This seems like a great idea! Adding @ilya-bobyr as he is really knowledgeable when it comes to this sort of thing.

The only real issue I see is that in CI, we check fmt/check/clippy against a specific nightly version (specified in `rust-version.sh` file). It would be ideal to run the pre-commit hooks against this same version

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      rev: v1.0
+      hooks:
+        # FIXME: enable fmt hook when it no longer emits warnings with
+        # stable build. Currently it requires nightly build.
+        # Until then, devs should run `cargo +nightly fmt` manually.
+        #- id: fmt
+        - id: clippy
+          stages: [pre-push]
+        - id: cargo-check
+          stages: [pre-push]
```
Comment (illia-bobyr) [inline .pre-commit-config.yaml]: We probably do not need to cater to everyone, but a lot of people create many commits that they expect GitHub to squash for them [^1].
I am a bit concerned that these will be too slow.

Our CI uses the following scripts:
1. `fmt`: `scripts/cargo-fmt.sh`
2. `clippy`: `scripts/cargo-clippy.sh`
3. I'm actually not sure the exact command to run all the tests.

But even the `fmt` script is quite slow.
If I need to run `fmt` myself, I always specify the affected package explicitly, as in:
```bash
./cargo nightly fmt --package solana-core
```

`clippy` is even slower.

And if you run the wrong version, you might get discrepancies from what you get later in CI.

[^1]: I personally prefer a different style of creating changes.
But, unfortunately, GitHub review process is more or less designed to work this way :(

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    # Language-agnostic hooks.
+    - repo: https://github.com/pre-commit/pre-commit-hooks
+      rev: v4.6.0
+      hooks:
+        - id: check-merge-conflict
+        - id: check-yaml
+        - id: check-toml
+        - id: end-of-file-fixer
+        - id: trailing-whitespace
```
Comment (illia-bobyr) [inline .pre-commit-config.yaml:61]: I think this is already not allowed by `git` by default, is it not?

**Comment by illia-bobyr** [comment]
> #### Problem
> 
> Manually running code checks before committing code or putting up a PR is tedious.

I think it would be nice to improve our "commit hygiene" a bit.
Though, to be fair, it is pretty good already.
We follow a policy of 1 commit per PR, so the CI will run `cargo fmt`, and `cargo clippy` and `cargo test`.
And if you do not submit without a green CI (which most people can not?), most of the checks will be enforced.

> #### Summary of Changes
> 
> * Add a yaml config for the pre-commit tool, so I can configure and install the appropriate GitHub hooks and automate running these checks on every code commit and push.
> 
> * Document tool usage and link to the relevant documentation.
> 
> * There is no impact to CI or developers that choose not to adopt this framework.
> 
> * Caveat: Cargo fmt is not enabled yet. The current setup I am leveraging uses cargo stable by default, and fmt requires nightly on this repo.

I wonder if using a tool like this is the best possible choice.
The rest of the CI is a set of bash scripts that can be called from other tools.
But now we would have this tool integrated in the process.

I do not think it would be wise not to call it from the CI.
We would want everyone to be on the same page.
For example, I do not want to deal with a trailing whitespace someone else has added.
If that person does not use the tool and I do, the issue will only pop up when I will try to run the checks.  Unless CI will also run the checks.

Now, as this tool runs checks on commit or push, there is a limit on which checks it can run.
Only those that are relatively quick can be included.
And that creates a partitioning for our CI checks.
Ultimately, it seems better if all checks are executed as part of the build process.

To be fair, our build system, `cargo`, is not flexible enough to easily include arbitrary checks.
So, maybe, this tool is a decent compromise short term.

**Comment by fkouteib** [comment]
Thanks Illia and Brennan for the review and feedback. All fair points. 

> I wonder if using a tool like this is the best possible choice. The rest of the CI is a set of bash scripts that can be called from other tools. But now we would have this tool integrated in the process.

I think we are trying to solve different problems causing the disconnect. I believe your focus and concern is having configured scripts that are manually run by devs and run in CI. That's great. I am trying to solve a different problem; My own problem of sometimes forgetting to run the non-test checks (i.e. formatting, whitespace, etc..), that I wouldn't normally do while iterating locally on my code, failing a CI run after I put up a PR for review, then iterating for those fixes. But your point is well taken on introducing another dependency and whether the ROI is there.

> I do not think it would be wise not to call it from the CI. We would want everyone to be on the same page.

I was going with "bottoms up" approach. Starting with volunteer opt-in, some devs try it. They (hopefully) like it and over time we can move to make it default opt-in. "top down" approach, we gotta get everyone bought in before they have had a chance to try it or see its value, and it's an uphill battle for something I quickly put together and didn't work out every last quirk and detail to satisfy a wider audience and a mandate. 

As for parity in CI, it can be run across all files, "$pre-commit run all-files". We would have to bring the repo to parity first then turn it on in CI (assuming all the other versioning and nightly vs stable cargo concerns are addressed, which I have a possible solution in mind for if we align on broad direction).  

> If that person does not use the tool and I do, the issue will only pop up when I will try to run the checks. Unless CI will also run the checks.

Your local runs, by default, only include files that you modified and staged. But yes within those files, it's either auto-fixed (i.e. whitespace; end of file) or it penalizes the early adopter which would be counter-productive.

> Ultimately, it seems better if all checks are executed as part of the build process.

Yes that seems like the better direction to go. One-button push (alas still manual trigger for dev, but automated for CI) expanding on the shell scripts we have today.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (fkouteib) [reply .pre-commit-config.yaml:61]: I am 99% sure git itself does not block nor fix trailing whitespace. What it does is offer the option to "ignore whitespace" in a diff. Various editors and IDEs have configs that can be configured to do it on file save though. 

By default, the checks on this list only run on locally modified and staged files (There is a check all files option we can use for CI). So it'd be cheaper and faster imo than asking every dev to find the config for their favorite editor and enable it to do this (assuming we adopt the tool on other merits).

**Comment by illia-bobyr** [comment]
> > I wonder if using a tool like this is the best possible choice. The rest of the CI is a set of bash scripts that can be called from other tools. But now we would have this tool integrated in the process.
> 
> I think we are trying to solve different problems causing the disconnect. I believe your focus and concern is having configured scripts that are manually run by devs and run in CI. That's great. I am trying to solve a different problem; My own problem of sometimes forgetting to run the non-test checks (i.e. formatting, whitespace, etc..), that I wouldn't normally do while iterating locally on my code, failing a CI run after I put up a PR for review, then iterating for those fixes. But your point is well taken on introducing another dependency and whether the ROI is there.

I do not think that adding a few more formatting, whitespace, and end of file consistency checks is bad.
Why not add them for everyone and make CI enforce it?
If this is a good tool, maybe we would want to express them using this tool.

I think that bash is the worst programming language (if you ignore those that are specifically designed to be bad).
The reason I mention bash scripts and CI is so that we consider consistency when making a decision here.
It is possible that in order to express all these checks we would want some tool.

> [...]
> 
> As for parity in CI, it can be run across all files, "$pre-commit run all-files". We would have to bring the repo to parity first then turn it on in CI (assuming all the other versioning and nightly vs stable cargo concerns are addressed, which I have a possible solution in mind for if we align on broad direction).

I do not think it is that hard.
[`tokei`](https://github.com/XAMPPRocky/tokei) says the validator repo is 800k lines total, with 683k lines of code.
512k is Rust, which, thanks to `fmt` and other Rust tools, is in a pretty good shape.

**Comment by fkouteib** [comment]
@ilya-bobyr I am aligned. I dropped the the rust hooks. I also added a secrets guardrail check. I ran pre-commit run --all-files locally. You are right. about 60 files impacted, and it's only trailing whitespace and EOF check. All auto-fixed by the tool run. So there is no process/tool issue from enabling it for devs and CI at the same time (once we have team buy-in).

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    - repo: https://github.com/pre-commit/pre-commit-hooks
+      rev: v4.6.0
+      hooks:
+        - id: check-merge-conflict # merge conflict string(s) found.
+        - id: check-yaml
+        - id: check-toml
+        - id: end-of-file-fixer
+        - id: trailing-whitespace
+        - id: detect-private-key
+        - id: check-added-large-files # Set up and use git lfs if this fails.
```
Comment (illia-bobyr) [inline .pre-commit-config.yaml:63]: Not sure if we are going to set up lfs.
Though, I also do not know what are the ranges here.

**Review by illia-bobyr — COMMENTED** [review]
I'm a bit worried this is bringing both Python and Go as dependencies.
Our dependency story is not very well defined, as far as I understand.
Other than that, it seems fine to me.

File: .pre-commit-config.yaml (hunk 3 above)
Comment (fkouteib) [reply .pre-commit-config.yaml:63]: The idea is we'd halt and have the conversation if a large file (or binary blob) makes sense to check in, and whether we'd want to bloat the repo metadata/history with it forever, set up git lbs, or adjust course and drop the large file altogether.

**Comment by fkouteib** [comment]
I solved the prior contention of needing to run specific versions of rust/cargo and nightly vs formal release for various commands using the "local repo" capability of pre-commit tool. However, the timing (i.e.. execution time and stall of git command) concern is unresolved with no clear path to resolution (clippy will take its nominal time to run across a large repo).

So I am shelving this work, and closing this PR.

**Pull request closed without merging**